### PR TITLE
Add support for STM32F7

### PR DESCRIPTION
**Pull request by dhylands**
Tested on the STM32F7 Discovery board.

### CODE REVIEW

File: src/hal/stm32f7/init.rs [hunk 1]
```diff
@@ -0,0 +1,348 @@
+// Zinc, the bare metal stack for rust.
+// Copyright 2014 Vladimir "farcaller" Pouzanov <farcaller@gmail.com>
```
Comment (farcaller) [inline src/hal/stm32f7/init.rs]: please update to state original authorship.

File: src/hal/stm32f7/timer.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  #[inline(always)]
+  fn get_counter(&self) -> u32 {
+    self.reg.CNT()
+  }
+}
+
+mod reg {
+  use volatile_cell::VolatileCell;
+
+  ioreg_old!(TIM2To5: u32, CR1, CR2, SMCR, DIER, SR, EGR, CCMR1, CCMR2, CCER, CNT,
```
Comment (farcaller) [inline src/hal/stm32f7/timer.rs]: can you convert this to new ioreg format?

File: src/hal/stm32f7/init.rs (hunk 1 above)
Comment (dhylands) [inline src/hal/stm32f7/init.rs]: I copied the stm32f4 file and made a few tweaks, so I can hardly claim authorship.

Comment (farcaller) [inline src/hal/stm32f7/init.rs]: Yep, I realised it later, sorry for the noise (came to be my go-to comment on these files).

**Comment by dhylands** [comment]
I've updated to use the new ioregs macro. If this looks good, then I'd be happy to backport to the stm32f4 (since it's virtually identical to the stm32f7).

There are a couple of issues I've come across using the new macros.

1 - It would be nice if there was also a get/set of the entire register even though there are lots of sub-fields.
2 - I haven't figured out a good way to deal with the timer CCMRx registers. These are really a union depending on whether the channel is configured as an input or output. For now, I've just left it as a 32-bit field.

**Comment by dhylands** [comment]
I rather like how the peripheral clocks are being dealt with in the stm32f1 port, so I think I'll convert the f7 port to do it that way.

**Comment by dhylands** [comment]
It looks like I'm going to have to work on something else, so I probably won't get a chance to finish up my peripheral clock changes in the near future, so I think that this is as much as will get done for the F7 for now. Please let me know if there are any other small changes you want made.

**Comment by farcaller** [comment]
Looks ok, thanks for working on this.

**Pull request merged**